Gems/Atom/RHI/Code/Include/Atom/RHI/MultiDeviceFence.h

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+#pragma once
+
+#include <Atom/RHI/Fence.h>
+#include <Atom/RHI/MultiDeviceObject.h>
+
+namespace AZ::RHI
+{
+    //! A multi-device synchronization primitive, holding device-specific Fences, than can be used to insert
+    //! dependencies between a queue and a host
+    class MultiDeviceFence : public MultiDeviceObject
+    {
+    public:
+        AZ_CLASS_ALLOCATOR(MultiDeviceFence, AZ::SystemAllocator, 0);
+        AZ_RTTI(MultiDeviceFence, "{5FF150A4-2C1E-4EC6-AE36-8EBD1CE22C31}", MultiDeviceObject);
+
+        MultiDeviceFence() = default;
+        virtual ~MultiDeviceFence() = default;
+
+        //! Initializes the multi-device fence using the provided deviceMask.
+        //! It creates on device-specific fence for each bit set in the deviceMask and
+        //! passes on the initial FenceState to each Fence
+        ResultCode Init(MultiDevice::DeviceMask deviceMask, FenceState initialState);
+
+        //! Waits on m_waitThread and shuts down all device-specific fences.
+        void Shutdown() override final;
+
+        //! Signals the device-specific fences managed by this class
+        RHI::ResultCode SignalOnCpu();
+
+        //! Waits (blocks) for all device-specific fences managed by this class
+        RHI::ResultCode WaitOnCpu() const;
+
+        //! Resets the device-specific fences.
+        RHI::ResultCode Reset();
+
+        using SignalCallback = AZStd::function<void()>;
+
+        //! Spawns a dedicated thread to wait on all device-specific fences. The provided callback
+        //! is invoked when the fences complete.
+        ResultCode WaitOnCpuAsync(SignalCallback callback);
+
+        //! Returns the device-specific Fence for the given index
+        const Ptr<Fence>& GetDeviceFence(int deviceIndex)
```

Comment:
We debated that as well, the reason we did not go down this route was mainly that currently virtually no other class in the RHI is templatized (because by moving the map to some parent class would require templatizing the MD-objects) and also most classes already have such repeated functionality (like `GetDescriptor` in some variation, also not templatized).

What makes templatizing a bit challenging is that not all MD objects have the same inheritance structure, i.e. `MultiDeviceBuffer` is a `MultiDeviceResource`, while `MultiDeviceFence` is only a `MultiDeviceObject` and others, like the *Items do not inherit at all.

But if this is preferred, I could quickly check and see if that would change a lot or would work fine.

Reply:
I'm not sure what the benefit of that is in comparison to having a template class? You would have to cast quite often (i.e., whenever you access the map) thought? I will try the template class approach and report back if I encounter any issues with it.

Reply:
Looks like the `AZ_RTTI` macro doesn't support a parent class with multiple template parameters (I need two for resource and pool type). So either that gets fixed (I don't think I can do that) or we will have to take another route. I'm not too fond of having to cast every time a map element is accessed :-/

Reply:
All *Item classes (Draw/Dispatch/DispatchRays), IndirectBufferWriter, PipelineLibraryDescriptor (see !16261) and ShaderResourceGroupData. I think it's best to stick with the maps in each object for now. It keeps the API as simple as possible without users of the API having to pass template parameters. The same would be possible with a template class, but for that we would need the `AZ_RTTI` macro to support multiple template parameters.

Reply:
Do you mean something like this? - https://github.com/o3de/o3de/blob/ec22e96a90392793025b6b961aa228e122612ff6/Code/Framework/AzCore/Tests/Components.cpp#L1972-L1980

Reply:
The Doxygen comments in the Component.h header also goes over how to implement AzTypeInfo, RTTI and the Component Descriptor for templated components [here](https://github.com/o3de/o3de/blob/ec22e96a90392793025b6b961aa228e122612ff6/Code/Framework/AzCore/AzCore/Component/Component.h#L452-L459)

Reply:
We have an example of using 1 template parameter in the Rtti but you should be bale to use it add another one [Unit Test](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/Tests/Rtti.cpp#L695-L719).